python/graphstorm/model/hgt_encoder.py

```diff
@@ -0,0 +1,288 @@
+"""
+    Copyright 2023 Contributors
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+    Heterogeneous Graph Transformer (HGT) layer implementation
+"""
+import warnings
+
+import math
+import torch
+import torch.nn.functional as F
+import dgl.function as fn
+
+from torch import nn
+from dgl.nn.functional import edge_softmax
+from .ngnn_mlp import NGNNMLP
+from .gnn_encoder_base import GraphConvEncoder
+
+
+class HGTLayer(nn.Module):
+    r"""Heterogenous graph transformer (HGT) layer.
+    
+    Different from DGL's HGTConv, this implementation is based on heterogeneous graph.
+    Other hyperparameters' default values are same as the DGL's HGTConv setting.
+    
+    Parameters
+    ----------
+    in_dim : int
+        Input dimension size.
+    out_dim : int
+        Output dimension size.
+    ntypes: list[str]
+        List of node types
+    canonical_etypes: list[(str, str, str)]
+        List of canonical edge types
+    num_heads : int
+        Number of attention heads
+    dropout : float, optional
+        Dropout rate. Default: 0.2
+    use_norm: boolean
+        If use layer normalization or not, default is True
+    num_ffn_layers_in_gnn: int, optional
+        Number of layers of ngnn between gnn layers
+    """
+    def __init__(self,
+                 in_dim,
+                 out_dim,
+                 ntypes,
+                 canonical_etypes,
+                 num_heads,
+                 dropout=0.2,
+                 use_norm=True,
+                 num_ffn_layers_in_gnn=0,
+                 fnn_activation=F.relu):
+        super(HGTLayer, self).__init__()
+        self.num_heads = num_heads
+        self.out_dim = out_dim
+        self.d_k = out_dim // num_heads
+        self.sqrt_dk = math.sqrt(self.d_k)
+        self.use_norm = use_norm
+
+        # Define node type parameters
+        k_linears = {}
+        q_linears = {}
+        v_linears = {}
+        a_linears = {}
+        norms = {}
+        skip = {}
+        for ntype in ntypes:
+            k_linears[ntype] = nn.Linear(in_dim, out_dim)
```

Comment:
How do you initialize k_liners, q_liners, etc?

Reply:
Based on this blog, they are initialized by Keiming/He algorithm.
https://jamesmccaffrey.wordpress.com/2023/02/07/understanding-the-pytorch-linear-layer-default-weight-and-bias-initialization/